### ISSUE TITLE
[2.2.6-rc1 bug] Existing datasets "Wrong Parameter" and "Permission Denied" error

### ISSUE THREAD

**Issue by thesn10** [openzfsonwindows/openzfs#397]
I have existing datasets, created in rc6, and i upgraded to 2.2.6-rc1 and cannot access them anymore.

Double clicking on a dataset folder gives permission error (that cannot be fixed by claiming permissions):

![image](https://github.com/user-attachments/assets/16e89dff-259b-4d9c-8afe-e101e5a01c3a)

Translated:
![image](https://github.com/user-attachments/assets/740716a4-e40f-416f-a14e-f6ee3225d0da)

Trying to `cd` into dataset gives "Invalid parameter" error:
```
Z:\>cd dataset
Invalid Parameter.
```
Using powershell, `cd` works but it shows the error after `dir`:
```
PS Z:\> cd dataset
PS Z:\dataset> dir
Get-ChildItem: Invalid Parameter. : 'Z:\dataset'
```

When using Filespy on the Explorer double click, there are multiple STATUS_INVALID_PARAMETER errors:
![image](https://github.com/user-attachments/assets/519db02c-55f1-42d4-94f3-6f6710a04a27)

When using Filespy on `cd`, it seems to fail at the first STATUS_REPASE:
![image](https://github.com/user-attachments/assets/80a43a53-6a95-41ef-9b5a-c5e6d949db86)

Also, zfs mount does only show one of my two datasets. I have a dataset named "dataset" and one named "userfiles". The latter does not even appear, and the other does not show the drive letter:
```
> zfs mount

tank                            Z:/
tank/dataset                    tank/dataset
```

zpool import tank output:

```
> zpool import tank
path '\\?\scsi#disk&ven_&prod_st16000nm000h-3k#5&1bc941f&0&000000#{53f56307-b6bf-11d0-94f2-00a0c91efb8b}'
 and '\\?\PhysicalDrive0'
read partitions ok 2
    gpt 0: type c4ccfd510 off 0x100000 len 0xe8d7f600000
    gpt 1: type c4ccfd510 off 0xe8d7f700000 len 0x800000
asking libefi to read label
EFI read OK, max partitions 9
    part 0:  offset 800:    len 746bfb000:    tag: 4    name: 'zfs-00003c1f00003374'
    part 8:  offset 746bfb800:    len 4000:    tag: b    name: ''
working on dev '#1048576#16000890175488#\\?\scsi#disk&ven_&prod_st16000nm000h-3k#5&1bc941f&0&000000#{53f56307-b6bf-11d0-94f2-00a0c91efb8b}'
setting path here '/dev/physicaldrive0'
setting physpath here '#1048576#16000890175488#\\?\scsi#disk&ven_&prod_st16000nm000h-3k#5&1bc941f&0&000000#{53f56307-b6bf-11d0-94f2-00a0c91efb8b}'
```

**Comment by thesn10** [comment]
Also unmount failed saying "tank is busy"

When i shutdown my pc, i got a BSOD saying
`PNP_DETECTED_FATAL_ERROR`

**Comment by lundman** [comment]
OK let's sum up, you have a `dataset` volume mount reparsepoint, which gives Security problems when you try to access it. If you were to create a new pool (temporary empty file pool is fine) does it happen? It is to answer if it is the Security stored on your pool, or, do we always make wrong Security.

2. The STATUS_INVALID is for method 37 - something we haven't implemented, and is most likely not the issue. Can you confirm with `mountvol` that the OS has everything mounted that you expect?

3. Tank is busy is a known issue, you can sometimes get around it by unmounting manually, one-by-one from deepest. Seems to be an issue with the "collected everything mounted" code.

**Comment by thesn10** [comment]
I got some more info about the BSOD i got.
The PNP_DETECTED_FATAL_ERROR has the following params:
```
Error code: 0x000000ca 
param1: 0x0000000000000004
param2: 0xffffe70a8f5a0e00
param3: 0x0000000000000000
param4: 0x0000000000000000
```

If i look at the [error code documentation](https://learn.microsoft.com/en-us/windows-hardware/drivers/debugger/bug-check-0xca--pnp-detected-fatal-error), we see our param1 is 0x4, which means that we enumerated a PDO which was previously deleted using `IoDeleteDevice`. So somewhere the zfs driver accesses a deleted PDO

![image](https://github.com/user-attachments/assets/f77ef6ad-5cf4-4d81-a307-51637fe03fb2)

**Comment by thesn10** [comment]
> OK let's sum up, you have a `dataset` volume mount reparsepoint, which gives Security problems when you try to access it. If you were to create a new pool (temporary empty file pool is fine) does it happen? It is to answer if it is the Security stored on your pool, or, do we always make wrong Security.
> 
> 2. The STATUS_INVALID is for method 37 - something we haven't implemented, and is most likely not the issue. Can you confirm with `mountvol` that the OS has everything mounted that you expect?
> 3. Tank is busy is a known issue, you can sometimes get around it by unmounting manually, one-by-one from deepest. Seems to be an issue with the "collected everything mounted" code.

Unfortunately I cannot create a new pool to test, as i dont have any spare drive. But i can show you the mountvol output:

mountvol: (Z:\ is the zpool)
```powershell
\\?\Volume{b2ebfafb-4cef-47ea-a357-c5ec53d88fcf}\
        E:\

    \\?\Volume{30c96a1e-1837-4960-b213-b0dd68e8def3}\
        C:\

    \\?\Volume{0bbb1c3d-1502-4a81-a103-2c7235749d5f}\
        *** NO MOUNT POINTS ***

    \\?\Volume{7fa68025-d442-43c7-9600-c502e106723c}\
        *** NO MOUNT POINTS ***

    \\?\Volume{ff3047ab-8539-44bd-8256-69dc4192e079}\
        *** NO MOUNT POINTS ***

    \\?\Volume{190625e6-6c32-11ef-8dda-ec086b081ef1}\
        G:\

    \\?\Volume{9b1c0123-30c0-11ef-8d81-ec086b081ef1}\
        Z:\

    \\?\Volume{9b1c056b-30c0-11ef-8d81-ec086b081ef1}\
    *** CANNOT BE PROVISIONED UNTIL A VOLUME MOUNT POINT HAS BEEN CREATED ***
```

So the Z:\dataset mountpoint is missing and it shows an error instead

`dir` shows completey different volume ids:
```powershell
PS Z:\> dir

    Directory: Z:\

Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
l----          23.06.2024    00:45                dataset -> Volume{1c14792a-a278-3dbe-a5ab-9e78981503d2}
d----          23.06.2024    02:42                folder
l----          22.06.2024    22:58                userfiles -> Volume{5447f2a6-1f62-36a3-9dad-c64a15afa84e}
```

The `wmiobject` does not show capacity or filesystem for zfs:

```powershell
Z:\> get-wmiobject Win32_Volume | ft @{Label = 'Mount'; Expression = {$_.Name -replace '^\\.+', ''}}, Capacity, FileSystem, Label, DeviceID

Mount      Capacity FileSystem Label        DeviceID
-----      -------- ---------- -----        --------
E:\   3790935683072 NTFS       4TB          \\?\Volume{b2ebfafb-4cef-47ea-a357-c5ec53d88fcf}\
C:\    499303575552 NTFS                    \\?\Volume{30c96a1e-1837-4960-b213-b0dd68e8def3}\
          628092928 NTFS                    \\?\Volume{0bbb1c3d-1502-4a81-a103-2c7235749d5f}\
          536854528 FAT32                   \\?\Volume{7fa68025-d442-43c7-9600-c502e106723c}\
          153092096 FAT32                   \\?\Volume{ff3047ab-8539-44bd-8256-69dc4192e079}\
G:\     16106127360 FAT32      Google Drive \\?\Volume{190625e6-6c32-11ef-8dda-ec086b081ef1}\
Z:\                                         \\?\Volume{9b1c0123-30c0-11ef-8d81-ec086b081ef1}\
                                            \\?\Volume{9b1c056b-30c0-11ef-8d81-ec086b081ef1}\
```

**Comment by guenther-alka** [comment]
Creating a testpool without physical disks
```
- open disk management (Datenträgerverwaltung) and create a virtual harddisk vhdx with dynamic size (real size depends on content)
- This creates and provides a virtual harddisk that you can use like a real disk (for a ZFS pool)
- If you use my napp-it ZFS web-gui you can manage Windows (Virtual HardDisks, Storage Spaces and ZFS) via web-GUI
```

**Comment by lundman** [comment]
Yeah, so you can just create a VHDX disk, even just a 1g, dynamic, will get you a PHYSICALDISKx to create a pool on. This is the Windows way. But you can also make a sparse file:
```
# fsutil file createnew C:\poolfile.bin 200000000
# zpool.exe create tank \\?\C:\poolfile.bin
```

but honestly, just make a vhdx.


But, the PNP at unload I have come across - so I can dig into here.

**Comment by guenther-alka** [comment]
The vhdx method (inherited from Hyper-V) is very flexible and very fast. It is a good idea to place all .vhdx files in the same folder ex c:\vhdx as you need to find and provide them as disks via a Powershell command after a reboot (can be a cmd file/planned task together with a pool import)

If you SMB share the folder, you can even provide a .vhdx as a disk over LAN. I have added this as a ZFS cluster method in my web-gui for a network mirror or Raid-Z even for Windows 11. With Windows Server methods like SMB direct/RDMA and nics > 10G, such a disk or SMB share over LAN is nearly as fast as a local disk even when using NVMe.

if ZFSonWindows becomes a little more stable I would say it has the power to be the most attractive ZFS plattform especially due SMB (direct) and ACL handling that is a pain with Linux and SAMBA and not available at all or not as fast.

**Comment by thesn10** [comment]
@lundman I have created a fresh pool and dataset and it works fine, no errors. 
But there seems to be a difference to migrated pools (from 2.2.3-rc6 to 2.2.6-rc1) in how the datasets are handled

Freshly created pool mounts fine:
```powershell
> mountvol

    \\?\Volume{80a3a563-6dcb-11ef-8ddc-ec086b081ef1}\
        D:\

    \\?\Volume{80a3a569-6dcb-11ef-8ddc-ec086b081ef1}\
        D:\dataset\
```

Migrated pool fails to mount dataset:
```powershell
> mountvol

    \\?\Volume{9b1c0123-30c0-11ef-8d81-ec086b081ef1}\
        Z:\

    \\?\Volume{9b1c056b-30c0-11ef-8d81-ec086b081ef1}\
    *** CANNOT BE PROVISIONED UNTIL A VOLUME MOUNT POINT HAS BEEN CREATED ***
```

- In the old migrated pool (Z:\) from 2.2.3-rc6 to 2.2.6-rc1, the datasets are displayed as folder junctions
- In the freshly created pool (D:\) in 2.2.6-rc1, the datasets are displayed as drives

![image](https://github.com/user-attachments/assets/b5658b97-bd0f-4174-a73f-a5ea1c121a3f)

![image](https://github.com/user-attachments/assets/1589731d-6dcc-40ad-ad2a-ee2a8335fa73)
![image](https://github.com/user-attachments/assets/fab5192d-acce-4d5e-9060-6bf2ec4b43fc)

**Comment by lundman** [comment]
Oh that is interesting. The main difference there is probably the property `driveletter` - a migrated pool wouldn't have the property at all. The code is supposed to pretend a non-existing `driveletter` to be the same as default, but maybe that goes wrong.

**Comment by lundman** [comment]
OK that seems to confirm a problem:
```
NAS$ mkfile -n 1g /var/tmp/pool
NAS$ zpool create DOOM /var/tmp/pool
NAS$ zfs create DOOM/dataset
NAS$ tar -cSf pool.tar /var/tmp/pool
NAS$ scp pool.tar vm:

vm$ tar -xSf pool.tar
vm$ zpool import -d `pwd` DOOM
vm$ zfs mount DOOM
cannot mount 'DOOM': Unknown error
```

OK, I'll need some time to debug this

**Comment by lundman** [comment]
OK so I straight up ignored SecurityDecriptor passed along with mkdir, and didn't merge with parent for when creating files. Please try rc4 and see if it has improved.

**Comment by thesn10** [comment]
@lundman 
> OK so I straight up ignored SecurityDecriptor passed along with mkdir, and didn't merge with parent for when creating files. Please try rc4 and see if it has improved.

### Installing rc4:

While installing rc4, i got a BSOD saying `SYSTEM_THREAD_EXCEPTION_NOT_HANDLED` in `OpenZFS.sys`. The exception code is  `STATUS_ACCESS_VIOLATION` (0xc0000005)

I retried installing it again and it was successfull.

### Testing mirgated pool on rc4:

It has improved a little bit on rc4: 

- The `dir` command on a dataset does not give an error anymore and lists files correctly now

But thats about it. The other errors have not improved:

- If i use explorer, the security error persists
- The recycle bin error persists
- The mountpoint is also still broken on rc4:

```powershell
> mountvol

    \\?\Volume{9b1c0123-30c0-11ef-8d81-ec086b081ef1}\
        Z:\

    \\?\Volume{9b1c056b-30c0-11ef-8d81-ec086b081ef1}\
    *** CANNOT BE PROVISIONED UNTIL A VOLUME MOUNT POINT HAS BEEN CREATED ***
```

### Setting driveletter again

> Oh that is interesting. The main difference there is probably the property driveletter - a migrated pool wouldn't have the property at all. The code is supposed to pretend a non-existing driveletter to be the same as default, but maybe that goes wrong.

I had set the driveletter to `Z:` on  2.2.3-rc6 but it seems to got lost in the migration:

```
> zfs get driveletter tank
bad property list: invalid property 'driveletter'
```

I tried setting the driveletter again and got this weird behaivior:

```
>zfs set driveletter=Z tank
>zpool export tank
>zpool import tank
>zfs mount
tank                            Z:/
tank/dataset                    //./Volume{0d99a5ba-7291-11ef-8de6-ec086b081ef1}//tank/dataset
tank/userfiles                  //./Volume{0d99a5ba-7291-11ef-8de6-ec086b081ef1}//tank/userfiles
```

Now my datasets are completely missing in file explorer

**Comment by lundman** [comment]
OK interesting. 

`bad property list: invalid property 'driveletter'`

Surprising - as if the property isnt defined in the code. I will check that.

Could you reboot, and "import -N tank" to stop it from mounting, then mount one at a time, starting with `zfs mount tank`. If it goes wrong, save cbuf for me. If it works, keep going...

`//./Volume{0d99a5b` ... just means mount failed.

**Comment by thesn10** [comment]
@lundman  I dont know how to save cbuf for a running zfs driver, as i am new to WinDbg. How do i attach to a running zfs driver? 
I am able to save cbuf for a bsod dump, but thats about it.

So, i rebooted and I executed:

```
zpool import -N tank
zfs mount tank
zfs mount tank/dataset
zfs mount tank/userfiles
```

No error, the commands where successful.
mountvol still shows invalid mounts.

- But strangely, for one dataset (named userfiles), it works and i can view the file content in explorer
- For the second dataset (named dataset), it doesnt work and i get the security error

I then executed `zpool export tank` and i got a BSOD.

Here is the bsod info:
[info.txt](https://github.com/user-attachments/files/17006308/info.txt)
[cbuf.txt](https://github.com/user-attachments/files/17006312/cbuf.txt)

This is the cbuf for the bsod after the `zpool export`, but how do i save the cbuf for a running zfs driver without a bsod?

**Comment by lundman** [comment]
Oh right, to save cbuf on running system: https://openzfsonosx.org/wiki/Windows_cbuf

but hold on for next release, has more mount/unmount fixes.

**Comment by lundman** [comment]
I added a tunable to ignore on disk SecurityDescriptor which may help this issue

**Comment by thesn10** [comment]
So I downloaded rc7 and  set the `windows_load_security` to 0

I have two datasets named "`userfiles`" and "`dataset`", the former works now, the latter does'nt.

- The dataset that does'nt work, does not show up in mountvol, and if i try to access it, i get the security error
- The other dataset works completely fine, which is strange, because they have the same configuration

I still get the recycle bin error

Here is the cbuf:

[cbuf.txt](https://github.com/user-attachments/files/17397298/cbuf.txt)

Also, Thanks for your work so far!

**Comment by thesn10** [comment]
Thanks for your work! Issue resolved in 2.26-rc9
